### ISSUE TITLE
Axum API: add timeout layer to avoid DoS attack

### ISSUE THREAD

**Issue by josecelano** [torrust/torrust-index-backend#204]
Parent issue: https://github.com/torrust/torrust-index-backend/issues/156

As described [here](https://github.com/torrust/torrust-tracker/issues/324), we need to set a timeout for requests to avoid a [denial of service attack](https://cheatsheetseries.owasp.org/cheatsheets/Denial_of_Service_Cheat_Sheet.html#rate-limiting).

How to do it:

- https://docs.rs/tower/0.4.13/tower/timeout/struct.TimeoutLayer.html
- https://docs.rs/axum/latest/axum/error_handling/index.html#applying-fallible-middleware

**Comment by programatik29** [comment]
Ideally `hyper` should support this feature rather than `axum` since supporting it in `axum` would require having its own server implementation. Hopefully this will be resolved with `hyper-1.0` release.

**Comment by da2ce7** [comment]
I agree, We should follow the issue with Hyper, and since users of our software need to use a https reverse proxy in any case, we document that they need to implement the connection timeouts there.Once we have implemented https, then we should take on the responsibility for timeouts since it is reasonable to use the software without a reverse proxy.Cam.On 20.06.2023, at 10:35, Eray Karatay ***@***.***> wrote:﻿
Ideally hyper should support this feature rather than axum since supporting it in axum would require having its own server implementation. Hopefully this will be resolved with hyper-1.0 release.

**Comment by jmkng** [comment]
Hello all,

I see this issue is still open, I assume you are still having trouble with the timeout middleware? Maybe I can take a look.

**Comment by josecelano** [comment]
> Hello all,
> 
> I see this issue is still open, I assume you are still having trouble with the timeout middleware? Maybe I can take a look.

Hi @jmkng it seems the best solution would be to fix it in [hyper 1.0](https://github.com/hyperium/hyper/issues/3178), but it can take long until they release the version 1.0. Maybe in the mean time we could implement [the solution](https://gist.github.com/programatik29/36d371c657392fd7f322e7342957b6d1) proposed by @programatik29. It was fixed on the [axum-server](https://github.com/programatik29/axum-server)](https://github.com/programatik29/axum-server/pull/39). Maybe we could do something similar and remove the patch once Hyper supports timeouts.

**Comment by josecelano** [comment]
[axum-server](https://github.com/programatik29/axum-server) was supposed to include the feature to set up a timeout for the TLS handshake. We have moved to [axum-server](https://github.com/programatik29/axum-server) in the [tracker](https://github.com/torrust/torrust-tracker). However, It seems that feature was removed. I've opened a new issue on the axum-server repo: https://github.com/programatik29/axum-server/issues/116.

**Comment by josecelano** [comment]
I've applied @programatik29 [patch on the tracker](https://github.com/torrust/torrust-tracker/pull/865). I will do the same here.

@programatik29 can we use your code? Is there any specific license we should include?

**Comment by josecelano** [comment]
I've just realized the [patch on the tracker](https://github.com/torrust/torrust-tracker/pull/865) does not work when you enable TSL. I had to comment on the TimeoutAcceptor to make the TSL configuration work. See https://github.com/torrust/torrust-index/pull/584. After merging that PR this problem will be fixed only when TSL is disabled (HTTP) but I will not work for HTTPs.

**Comment by josecelano** [comment]
I've converted the discussion in the Axun repo into a issue: https://github.com/tokio-rs/axum/issues/2741

**Comment by josecelano** [comment]
A [PR](https://github.com/hyperium/hyper/pull/3185) has been merged in the [hyper repo](https://github.com/hyperium/hyper/pull/3185). It changes the `http1_header_read_timeout` timeout.

![image](https://github.com/torrust/torrust-tracker/assets/58816/5326e0a8-030e-47fb-8bb5-583177d82985)

This could fix this issue. We can try when this change is published in a new release.

**Comment by josecelano** [comment]
> A [PR](https://github.com/hyperium/hyper/pull/3185) has been merged in the [hyper repo](https://github.com/hyperium/hyper/pull/3185). It changes the `http1_header_read_timeout` timeout.
> 
> ![image](https://private-user-images.githubusercontent.com/58816/337560832-5326e0a8-030e-47fb-8bb5-583177d82985.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxNzUzMjksIm5iZiI6MTcyMDE3NTAyOSwicGF0aCI6Ii81ODgxNi8zMzc1NjA4MzItNTMyNmUwYTgtMDMwZS00N2ZiLThiYjUtNTgzMTc3ZDgyOTg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA1VDEwMjM0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE3M2I4NjA1NDI0NWUxNTc3M2I3MWNhYjQxYjI4YzI2Nzc1ODJhMTcwN2Q2MWMzOGVlZGRkMjVmNWU0YmRiNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.y2SIfApKsXR09pxqMrQ7s5kwXgG2BWqJAoMu-uN_YDI)
> 
> This could fix this issue. We can try when this change is published in a new release.

`hyper` 1.4.0 has been released with server starting header read timeout immediately (#3185) ([0eb1b6cf](https://github.com/hyperium/hyper/commit/0eb1b6cf4d914ce9c3f8e92a8b43754eba27a327))

- https://github.com/hyperium/hyper/blob/master/CHANGELOG.md#v140-2024-07-01
- https://github.com/torrust/torrust-tracker/pull/945
